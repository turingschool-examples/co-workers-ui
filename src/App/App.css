* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

main {
  display: flex;
  width: 100%;
}
