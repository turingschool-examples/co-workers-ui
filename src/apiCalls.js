export const getCoWorkers = async () => {
  const url = 'http://localhost:3001/api/v1/coworkers'
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting your co-workers.');
  }
  const coWorkers = await response.json();
  return coWorkers
}

export const createCoWorker = async newCoWorker => {
  const url = 'http://localhost:3001/api/v1/coworkers';
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ ...newCoWorker })
  };
  const response = await fetch(url, options);
  if (!response.ok) {
    throw new Error('Sorry.  Unable to create your new co-worker.');
  }
  const newCoWorkerId = await response.json();
  return newCoWorkerId;
}

export const deleteCoWorker = async id => {
  const url = `http://localhost:3001/api/v1/coworkers/${id}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);
  if (!response.ok) {
    throw new Error('There was a problem removing this co-worker.')
  }
}