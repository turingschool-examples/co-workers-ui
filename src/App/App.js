import React, { Component } from 'react';
import Form from '../Form/Form';
import Dashboard from '../Dashboard/Dashboard';
import Profile from '../Profile/Profile';
import { getCoWorkers, createCoWorker, deleteCoWorker } from '../apiCalls';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      coWorkers: [],
      selectedId: null,
      error: ''
    }
  }

  async componentDidMount() {
    try {
      const coWorkers = await getCoWorkers();
      this.setState({ coWorkers });
    } catch({ message }) {
      this.setState({ error: message });
    }
  }

  addCoWorker = async newCoWorker => {
    try {
      await createCoWorker(newCoWorker);
      const coWorkers = await getCoWorkers();
      this.setState({ coWorkers });
    } catch({ message }) {
      this.setState({ error: message });
    }
  }

  removeCoWorker = async (e, id) => {
    e.stopPropagation();
    try {
      await deleteCoWorker(id);
      const coWorkers = await getCoWorkers();
      this.setState({ coWorkers, selectedId: null });
    } catch({ message }) {
      this.setState({ error: message });
    }
  }

  selectUser = id => {
    this.setState({ selectedId: id});
  }

  render() {
    const { coWorkers, selectedId, error } = this.state;
    const foundUser = coWorkers.find(coWorker => coWorker.id === selectedId);
    return (
      <div className="app">
          <Form addCoWorker={this.addCoWorker} />
        <main>
          <Dashboard 
            coWorkers={coWorkers} 
            error={error} 
            removeCoWorker={this.removeCoWorker} 
            selectUser={this.selectUser}
          />
        <Profile selected={foundUser} />
        </main>
      </div>
    );
  }
}

export default App;
