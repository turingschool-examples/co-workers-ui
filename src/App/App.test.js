import React from 'react';
import { shallow } from 'enzyme';
import { App } from './App';
import { fetchCoWorkers, postCoWorker, deleteCoWorker } from '../apiCalls';

jest.mock('../apiCalls.js');

describe('App', () => {
  let wrapper
  const mockSetCoWorkers = jest.fn();
  const mockUpdateLoading = jest.fn();
  const mockHasErrored = jest.fn();
  const mockCoWorkers = [
    {
      id: 1,
      image: 'https://image.unsplash.com/1',
      location: 'Pennsylvania',
      name: 'Miguel',
      role: 'Owner',
      status: false
    },
    {
      id: 2,
      image: 'https://image.unsplash.com/2',
      location: 'Colorado',
      name: 'Hailie',
      role: 'Teammate',
      status: true
    }
  ]

  fetchCoWorkers.mockImplementation(() => Promise.resolve(mockCoWorkers));
  postCoWorker.mockImplementation(() => Promise.resolve({ id: Date.now() }));
  deleteCoWorker.mockImplementation(() => Promise.resolve([mockCoWorkers[0]]));

  beforeEach(() => {
    wrapper = shallow(<App 
      selectedId={1}
      isLoading={false}
      coWorkers={mockCoWorkers}
      errorMsg={''}
      setCoWorkers={mockSetCoWorkers}
      updateLoading={mockUpdateLoading}
      hasErrored={mockHasErrored}
    />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update loading, fetch coworkers, and setCoWorkers after mounting', () => {
    expect(mockUpdateLoading).toHaveBeenCalledWith(true);
    expect(fetchCoWorkers).toHaveBeenCalled();
    expect(mockUpdateLoading).toHaveBeenCalledWith(false);
    expect(mockSetCoWorkers).toHaveBeenCalledWith(mockCoWorkers);
  });

  it('should update loading, post a coWorker, fetch coWorkers, and setCoWorkers when addCoWorker is called', async () => {
    const mockCoWorker = {
      id: 3,
      image: 'https://images.unsplash.com/3',
      location: 'Colorado',
      name: 'Travis Rollins',
      role: 'Owner',
      status: true
    };
    await wrapper.instance().addCoWorker(mockCoWorker)

    expect(mockUpdateLoading).toHaveBeenCalledWith(true);
    expect(postCoWorker).toHaveBeenCalledWith(mockCoWorker);
    expect(fetchCoWorkers).toHaveBeenCalled();
    expect(mockUpdateLoading).toHaveBeenCalledWith(false);
    expect(mockSetCoWorkers).toHaveBeenCalledWith(mockCoWorkers);
  });

  it('should update loading, delete a coWorker, and setCoWorkers when removeCoWorker is called', async () => {
    const mockEvent = { stopPropagation: jest.fn() };

    await wrapper.instance().removeCoWorker(mockEvent, 2);

    expect(mockUpdateLoading).toHaveBeenCalledWith(true);
    expect(deleteCoWorker).toHaveBeenCalledWith(2);
    expect(mockUpdateLoading).toHaveBeenCalledWith(false);
    expect(mockSetCoWorkers).toHaveBeenCalledWith([mockCoWorkers[0]]);
  })
});
