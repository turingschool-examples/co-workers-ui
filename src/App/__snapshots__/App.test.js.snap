// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="app"
>
  <Form
    addCoWorker={[Function]}
  />
  <main>
    <Dashboard
      coWorkers={
        Array [
          Object {
            "id": 1,
            "image": "https://image.unsplash.com/1",
            "location": "Pennsylvania",
            "name": "Miguel",
            "role": "Owner",
            "status": false,
          },
          Object {
            "id": 2,
            "image": "https://image.unsplash.com/2",
            "location": "Colorado",
            "name": "Hailie",
            "role": "Teammate",
            "status": true,
          },
        ]
      }
      error=""
      removeCoWorker={[Function]}
    />
    <Profile
      selected={
        Object {
          "id": 1,
          "image": "https://image.unsplash.com/1",
          "location": "Pennsylvania",
          "name": "Miguel",
          "role": "Owner",
          "status": false,
        }
      }
    />
  </main>
</div>
`;
